core/src/main/scala/org/broadinstitute/dsp/workbench/welder/BackgroundTask.scala

```diff
@@ -116,18 +103,54 @@ class BackgroundTask(
-  def shouldDelocalize(gsPath: GsPath, localAbsolutePath: Path): IO[Boolean] =
+  def safeDelocalize(gsPath: GsPath, localObjectPath: RelativePath)(implicit ev: Ask[IO, TraceId]): IO[Unit] =
     for {
-      meta <- googleStorageService.getObjectMetadata(gsPath.bucketName, gsPath.blobName, None).compile.last
-      localCrc32c <- Crc32c.calculateCrc32ForFile(localAbsolutePath, blocker)
-    } yield meta match {
-      case Some(GetMetadataResponse.Metadata(crc32, _, _)) =>
-        if (localCrc32c == crc32)
-          false
-        else
-          true
-      case _ => true
-    }
+      traceId <- ev.ask[TraceId]
+      localAbsolutePath = config.workingDirectory.resolve(localObjectPath.asPath) //TODO: make sure this is correct
+      _ <- logger.info(s"local absolute path ${localAbsolutePath}")
+      previousMeta <- metadataCacheAlg.getCache(localObjectPath)
+      calculatedCrc32c <- Crc32c.calculateCrc32ForFile(localAbsolutePath, blocker)
+
+      _ <- previousMeta match {
+        case Some(meta) =>
+          meta.remoteState match {
+            case RemoteState.NotFound =>
+              delocalizeAndUpdateCache(localObjectPath, gsPath, 0L, traceId)
+            case RemoteState.Found(_, crc32c) =>
+              if (calculatedCrc32c == crc32c)
+                IO.unit
+              else
+                delocalizeAndUpdateCache(localObjectPath, gsPath, meta.localFileGeneration.getOrElse(0L), traceId)
+          }
+        case None =>
+          delocalizeAndUpdateCache(localObjectPath, gsPath, 0L, traceId)
+      }
+    } yield ()
+
+  private def delocalizeAndUpdateCache(localObjectPath: RelativePath, gsPath: GsPath, generation: Long, traceId: TraceId): IO[Unit] = {
+    val hashedOwnerEmail = IO.fromEither(hashString(config.ownerEmail.value))
+    val lastModifiedByMetadataToPush: Map[String, String] = Map("lastModifiedBy" -> hashedOwnerEmail.toString())
+    for {
+      delocalizeResp <- googleStorageAlg
+        .delocalize(localObjectPath, gsPath, generation, lastModifiedByMetadataToPush, traceId)
+        .recoverWith {
+          case e: com.google.cloud.storage.StorageException if e.getCode == 412 =>
+            if (generation == 0L)
+              IO.raiseError(e)
+            else {
+              // In the case when the file is already been deleted from GCS, we try to delocalize the file with generation being 0L
+              // This assumes the business logic we want is always to recreate files that have been deleted from GCS by other users.
+              // If the file is indeed out of sync with remote, both delocalize attempts will fail due to generation mismatch
+              googleStorageAlg.delocalize(localObjectPath, gsPath, 0L, lastModifiedByMetadataToPush, traceId)
+            }
+        }
+        .recoverWith {
+          case e: com.google.cloud.storage.StorageException if e.getCode == 412 =>
+            googleStorageAlg.delocalize(localObjectPath, gsPath, 0L, Map("ownerEmail" -> "outdated"), traceId)
```

Comment:
I don't think we should delocalize again. This will just fail again with 412...We need to make a call that just updates metadata...also `ownerEmail` needs to be populated properly